### PR TITLE
Alternative osu!catch autoplay implementation

### LINKED ISSUE #3623: osu!catch autoplay breaks on maps with a large amount of objects
The autoplay feature on osu!catch seems to break on maps with large amounts of objects (not confirmed, have only tested on one map)

I know this was meant to be a visualization map, but this bug could happen elsewhere, so I figured I should report it.

OS: Manjaro Linux (kernel 4.18.12-1-MANJARO)
osu!lazer version (manually built by me): 2018.1017.0
Beatmap: https://osu.ppy.sh/beatmapsets/150945

Steps to reproduce:

- open osu!lazer
- select osu!catch gamemode
- click play
- click solo
- select centipede beatmap
- go to mods and turn on autoplay
- play beatmap

It should crash right after the first "spinner" (don't know if they're called anything different in osu!catch, i typically only play the normal osu! gamemode)

Here was the error I got:

```
Unhandled Exception: System.ArgumentException: X must be finite, but is ∞.
   at osu.Framework.Graphics.Drawable.set_X(Single value)
   at osu.Game.Rulesets.Catch.UI.CatcherArea.UpdateAfterChildren() in /home/joyesh/osu/osu.Game.Rulesets.Catch/UI/CatcherArea.cs:line 104
   at osu.Framework.Graphics.Containers.CompositeDrawable.UpdateSubTree()
   at osu.Framework.Graphics.Containers.CompositeDrawable.UpdateSubTree()
   at osu.Framework.Graphics.Containers.CompositeDrawable.UpdateSubTree()
   at osu.Framework.Graphics.Containers.CompositeDrawable.UpdateSubTree()
   at osu.Framework.Graphics.Containers.CompositeDrawable.UpdateSubTree()
   at osu.Framework.Graphics.Containers.CompositeDrawable.UpdateSubTree()
   at osu.Framework.Graphics.Containers.CompositeDrawable.UpdateSubTree()
   at osu.Game.Rulesets.UI.RulesetInputManager`1.UpdateSubTree() in /home/joyesh/osu/osu.Game/Rulesets/UI/RulesetInputManager.cs:line 141
   at osu.Framework.Graphics.Containers.CompositeDrawable.UpdateSubTree()
   at osu.Framework.Graphics.Containers.CompositeDrawable.UpdateSubTree()
   at osu.Framework.Graphics.Containers.CompositeDrawable.UpdateSubTree()
   at osu.Framework.Graphics.Containers.CompositeDrawable.UpdateSubTree()
   at osu.Framework.Graphics.Containers.CompositeDrawable.UpdateSubTree()
   at osu.Framework.Graphics.Containers.CompositeDrawable.UpdateSubTree()
   at osu.Framework.Graphics.Containers.CompositeDrawable.UpdateSubTree()
   at osu.Framework.Graphics.Containers.CompositeDrawable.UpdateSubTree()
   at osu.Framework.Graphics.Containers.CompositeDrawable.UpdateSubTree()
   at osu.Framework.Graphics.Containers.CompositeDrawable.UpdateSubTree()
   at osu.Framework.Graphics.Containers.CompositeDrawable.UpdateSubTree()
   at osu.Framework.Graphics.Containers.CompositeDrawable.UpdateSubTree()
   at osu.Framework.Graphics.Containers.CompositeDrawable.UpdateSubTree()
   at osu.Framework.Graphics.Containers.CompositeDrawable.UpdateSubTree()
   at osu.Framework.Graphics.Containers.CompositeDrawable.UpdateSubTree()
   at osu.Framework.Graphics.Containers.CompositeDrawable.UpdateSubTree()
   at osu.Framework.Graphics.Containers.CompositeDrawable.UpdateSubTree()
   at osu.Framework.Graphics.Containers.CompositeDrawable.UpdateSubTree()
   at osu.Framework.Graphics.Containers.CompositeDrawable.UpdateSubTree()
   at osu.Framework.Graphics.Containers.CompositeDrawable.UpdateSubTree()
   at osu.Framework.Graphics.Containers.CompositeDrawable.UpdateSubTree()
   at osu.Framework.Platform.GameHost.UpdateFrame()
   at osu.Framework.Threading.GameThread.ProcessFrame()
   at osu.Framework.Threading.GameThread.runWork()
--- End of stack trace from previous location where exception was thrown ---
   at osu.Framework.Threading.Scheduler.Update()
   at osu.Framework.Threading.GameThread.ProcessFrame()
   at osu.Framework.Platform.GameHost.<>c__DisplayClass98_0.<Run>b__1(Object <p0>, FrameEventArgs <p1>)
   at OpenTK.GameWindow.RaiseUpdateFrame(Stopwatch watch, Double elapsed, Double& timestamp)
   at OpenTK.GameWindow.DispatchUpdateFrame(Stopwatch watch)
   at OpenTK.GameWindow.Run(Double updates_per_second, Double frames_per_second)
   at osu.Framework.Platform.GameHost.Run(Game game)
   at osu.Desktop.Program.Main(String[] args) in /home/joyesh/osu/osu.Desktop/Program.cs:line 46
Aborted (core dumped)

### PR DESCRIPTION
**Pull request by Hanamuke**
Key differences with PR #3518 autoplay : 
- It will not break the rules (and go faster than possible)
- It will catch the maximum number of bananas possible within the rules.
- It handles stacked objects correctly
- It is way more consistent, since it takes better/slower trajectory
- The code is way more complex. 
- Fixes #3623.

The changes are especially noticeable on aspire-like maps (https://osu.ppy.sh/beatmapsets/490223#fruits/1044929) and on spinners : 
https://youtu.be/zPqNj73jilE

Since the code is quite long and mathematical (and therefore prone to errors and a pain to review), I thought it would make it easier to implement it as an alternative autoplay mod. At least it doesn't break anything. 

To put it simply, the idea of the algorithm is to compute for every Fruit StartTime, a function of the maximal score you can still get depending on your horizontal position. The MathUtils object CatchStepFunction has been implemented to fulfil that role. 

My idea is then to use this patch to code a mod that hides suboptimal bananas, to help train spinning.
I also have ideas of using similar concepts as additional hindsight for SR calculation.

Just tell me if this is uninteresting I mostly coded it as a challenge. A lot of players would like that mod though.

### CODE REVIEW

File: osu.Game.Rulesets.Catch/Replays/CatchAutoGenerator2.cs [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+                if (obj.StartTime != times[0])
+                {
+                    scores.Insert(0, new CatchStepFunction(scores[0], (float)(dash_speed * (times[0] - obj.StartTime))));
+                    times.Insert(0, obj.StartTime);
+                }
+                if (obj.HyperDash)
+                {
+                    float distance = Math.Abs(obj.HyperDashTarget.X - obj.X);
+                    scores[0].Set(Math.Max(0, obj.X - halfCatcherWidth), Math.Min(1, obj.X + halfCatcherWidth),
+                    scores[1].Max(obj.X - distance, obj.X + distance));
```
Comment (ekrctb) [inline osu.Game.Rulesets.Catch/Replays/CatchAutoGenerator2.cs]: Probably a mistake because this statement has no side-effect.

Comment (ekrctb) [inline osu.Game.Rulesets.Catch/Replays/CatchAutoGenerator2.cs]: Should indent because the two lines seems two separate statements.

**Comment by Hanamuke** [comment]
@ekrctb It's been almost three years, and I'll be super busy in the coming month. I will have some time this week-end, so I'll try to assess what needs to be done and try to fix integration. At the very least, I'll update you on sunday, and I don't mind you taking over if needs be.

**Comment by ekrctb** [comment]
@Hanamuke I have already started some more changes <https://github.com/ppy/osu/compare/master...ekrctb:auto2-wip>, and  I'm now planning to rewriting the code and opening a new PR, with an attribution to you for the algorithm.
If you are busy, don't worry about this PR :) This PR can be closed unless you want to push it to be merged by yourself.
Thank you for your contribution.

**Comment by Hanamuke** [comment]
@ekrctb Sorry for dumping that on you, that's quite a lot of work you've done. 

Testing locally my branch, I've noticed that it crashes on https://osu.ppy.sh/beatmapsets/90935/#fruits/265177, so you might wanna test it on your end also. I don't really have an idea why ; it is an avoid map so it has a lot of stacked and simultaneous notes, but the algorithm works fine on the hexagone avoid maps for instance. Speaking of avoid maps, it would be easy to make an autoavoid mod but simply using negative values for catching objects, if we want to do that down the line.
